LabWork №3 Ustinov

UAlex322 opened the pull request:
Основная часть готова. Остаётся прикрепить отчёт, добавить, возможно, математические функции и поправить недочёты.

UAlex322 committed:
Л/р №3, пока без отчёта

UAlex322 committed:
Перекодированные файлы

Теперь русский язык отображается нормально.

UAlex322 commented:
Сейчас заметил, что ещё нужно выводить виды ошибок в выражении. Исправлю.

UAlex322 committed:
Добавлен вывод ошибок в выражении

ArisovaAnastasia commented:
Есть некоторые замечания. 
Есть примеры, которые работают некорректно: 5--2 или 4/-2  Эти выражения не противоречат законам математики, поэтому считаются корректными и должны вычисляться.  
Также нужно добавить возможность, чтобы в качестве операндов могли быть переменные. Это требование указано в условии работы.

UAlex322 commented:
Хорошо, обработку таких выражений добавлю. Не совсем понятно, как вводить и использовать переменные. Можете, пожалуйста, привести какой-нибудь содержательный пример?

ArisovaAnastasia commented:
Например, пользователь ввел выражение 5*х, оно переводится в ОПЗ, при его подсчете у пользователя запрашивается значение переменной x и для него считается. Также желательно добавить возможность вычисления этого же выражения, но уже для другого значения х

UAlex322 commented:
Теперь ясно, спасибо.

UAlex322 committed:
Внесены многочисленные исправления

Добавлены работа с переменными, обработка не учтённых ранее выражений (пустое, с пробелами, с унарным минусом без скобок и др.), обработка неправильного ввода значений переменных. Парсинг и вычисление теперь происходят через класс Expression, который в конструкторе создаёт вектор лексем в обратной польской записи и потом позволяет неоднократно вычислять выражение, в том числе с подстановкой разных значений переменных, через метод compute().

UAlex322 commented:
Ещё могу попробовать добавить возведение в степень и функции, а также возможность добавлять цифры в названия переменных. Наверняка что-то упустил из виду, но все тесты парсер проходит.

ArisovaAnastasia commented:
Есть некоторые замечания:
1.  При повторном задании значения переменной, почему-то сначала выводится сообщение : "x= Incorrect input, try again" и только потом предлагается ввод значения : "x=". Это не совсем правильно, пользователь еще ничего не ввел, а ему сообщают что он ввел что-то некорректное.
2. Так при выходе из режима работы с одним выражением, печатается странная конструкция:

Enter the expr and press Enter (enter 'exit' to exit): Expression is incorrect!
Empty expression

И только потом пользователю предлагается ввести новое выражение. 
3. Данные примеры должны вычисляться, однако программа выдает сообщение, что они некорректны:

1/-(4+6)
1/--4  , но 1/-4 работает верно
1---1   , но  1--1 работает верно

ArisovaAnastasia commented:
Добавление возможности возведения в степень и стандартных функций опционально по вашему желанию

UAlex322 commented:
Ясно. А выражения с переменяющимися знаками "+" и "-" вроде "1++-+--+-4" тоже должны считаться правильными?

ArisovaAnastasia commented:
Нет, все бинарные операции должны быть справа ограничены операндами, заданными или вычисленными. В задании не требуется реализовать унарный '+' , требуется реализовать только унарный '-', поэтому верны будут только выражения вида 4---4 или 4+-4.

UAlex322 committed:
Внесены небольшие изменения

Сделаны заготовки под обработку правоассоциативных операций (унарные, возведение в степень). Унарный минус теперь работает правильно со скобками и множественным повторением.

ArisovaAnastasia commented:
Лабораторная работа принята

UAlex322 commented:
[Report LW №3 АиСД Устинов.docx](https://github.com/ArisovaAnastasia/mp2-lab3-arithmetic/files/7737305/Report.LW.3.docx)
Надеюсь, отчёт нормально отобразился. Я немного смешал описание программной реализации и метод решения, просто не знаю, как лучше там разграничить.

ArisovaAnastasia commented:
[Report.LW.3.docx](https://github.com/ArisovaAnastasia/mp2-lab3-arithmetic/files/7738440/Report.LW.3.docx)
Замечания к отчету

UAlex322 commented:
[Report LW №3 АиСД Устинов.docx](https://github.com/ArisovaAnastasia/mp2-lab3-arithmetic/files/7738893/Report.LW.3.docx)
Исправления внесены.

ArisovaAnastasia commented:
Отчет принят
